pulsar-broker-auth-oidc/src/main/java/org/apache/pulsar/broker/authentication/oidc/FallbackDiscoveryMode.java

```diff
@@ -0,0 +1,61 @@
+ */
+package org.apache.pulsar.broker.authentication.oidc;
+
+import org.apache.pulsar.common.classification.InterfaceStability;
+
+/**
+ * These are the modes available for configuring how the Open ID Connect Authentication Provider should handle a JWT
+ * that has an issuer that is not explicitly in the allowed issuers set configured by
+ * {@link AuthenticationProviderOpenID#ALLOWED_TOKEN_ISSUERS}. The current implementations rely on using the Kubernetes
+ * Api Server's Open ID Connect features to discover an additional issuer or additional public keys to trust. See the
+ * Kubernetes documentation for more information on how Service Accounts can integrate with Open ID Connect.
+ * https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-issuer-discovery
+ */
+@InterfaceStability.Evolving
+public enum FallbackDiscoveryMode {
+    /**
+     * There will be no discovery of additional trusted issuers or public keys. This setting requires that operators
+     * explicitly allow all issuers that will be trusted. For the Kubernetes Service Account Token Projections to work,
+     * the operator must explicitly trust the issuer on the token's "iss" claim. This is the default setting because it
+     * is the only mode that explicitly follows the OIDC spec for verification of discovered provider configuration.
+     */
+    DISABLED,
+
+    /**
+     * The Kubernetes Api Server will be used to discover an additional trusted issuer by getting the issuer at the
+     * Api Server's /.well-known/openid-configuration endpoint, verifying that issuer matches the "iss" claim on the
+     * supplied token, then treating that issuer as a trusted issuer by discovering the jwks_uri via that issuer's
+     * /.well-known/openid-configuration endpoint. This mode can be helpful in EKS environments where the Api Server's
+     * public keys served at the /openid/v1/jwks endpoint are not the same as the public keys served at the issuer's
+     * jwks_uri. It fails to be OIDC compliant because the URL used to discover the provider configuration is not the
+     * same as the issuer claim on the token.
+     */
+    KUBERNETES_DISCOVER_TRUSTED_ISSUER,
+
+    /**
+     * The Kubernetes Api Server will be used to discover an additional set of valid public keys by getting the issuer
+     * at the Api Server's /.well-known/openid-configuration endpoint, verifying that issuer matches the "iss" claim on
+     * the supplied token, then calling the Api Server endpoint to get the public keys using a kubernetes client. This
+     * mode is currently useful getting the public keys from the Api Server because the Api Server requires custom TLS
+     * and authentication, and the kubernetes client automatically handles those. It fails to be OIDC compliant because
+     * the URL used to discover the provider configuration is not the same as the issuer claim on the token.
+     */
+    KUBERNETES_DISCOVER_PUBLIC_KEYS,
```

Comment:
> It seems too laborious for a deployer to know which variant to choose between these two.

I plan to provide clear documentation explaining which to choose which setting. Because we are diverging from the OIDC discovery spec, I think it is important to provide users with information on the implementation so they can evaluate if the design meets their needs.

Essentially, it'll be a flow chart for how your k8s API Server works. EKS requires `KUBERNETES_DISCOVER_TRUSTED_ISSUER` right now, AKS requires `KUBERNETES_DISCOVER_PUBLIC_KEYS`, and GKE can do either. AWS does not work for `KUBERNETES_DISCOVER_PUBLIC_KEYS` because the API server does not serve the public keys.

It is possible that AWS will eventually support serving the public keys, so I don't want to make the configuration mode EKS specific. However, since the docs can easily be updated, I'll document it for now and then we can change it as the relevant options change.

> Frankly I don't recall in which situation you'd need to use `KUBERNETES_DISCOVER_PUBLIC_KEYS`.

This is needed when we're getting the public keys from the API Server. As we discussed off line, you could avoid some of this by letting users configure their own authentication and trusted ca cert for an issuer. However, I don't see any other implementation that requires authentication to retrieve the discovery doc or the jwks. It looks like k8s only requires authentication for the OIDC endpoints because of other issues they faced: https://github.com/kubernetes/kubernetes/pull/80724#issuecomment-577957929.